File: APC/my_codes/fv/14030225/14030225 APC-GPU/APC Matlab/general_APC.m
function Recived_profile = general_APC(s, y_noisy, N, alpha, sigma)

M = length(alpha); % Number of Iteration
y_noisy = [zeros(M*(N-1),1);y_noisy;zeros(M*(N-1),1)];

%%%%%%%%%% Calculate Henkel Matrix 'S' %%%%%%%%%%
temp = s;
S = zeros(N,2*N-1);
for i = N:2*N-1
    S(:,i) = temp;
    temp = [0;temp(1:end-1)];
end
temp = s;
for i = N-1:-1:1
    temp = [temp(2:end);0];
    S(:,i) = temp;
end

%%%%%%%%%% Calculate Initial Stage %%%%%%%%%%
sum_s = zeros(N,N);
for i=1:2*N-1
    sum_s = sum_s + S(:,i)*S(:,i)';
end
W_int = sum_s^-1*s; % Initial W(l)

%%%%%%%%%% Calculate X1(l) %%%%%%%%%%
temp_X = zeros(1,length(y_noisy)-(N-1));
for i= 1:length(y_noisy)-(N-1)
    temp_X(i) = W_int'*y_noisy(i:i+N-1);
end

 X = temp_X; % Estimated X
%  X = conv(s,y_noisy); % Estimated X
%  X=X(1:285);
%%%%%%%%%% Reiterative Algorithm Part %%%%%%%%%%%
for j=1:M
    temp_X = zeros(1,length(X));
    rho = abs(X).^alpha(j); % Calculate |p(l)|^2
    for i = N:length(X)-N+1
        % Calculation C(l)
        temp_rho = rho(i-N+1:i+N-1);
        C = zeros(N);
        for k=1:2*N-1
            C = temp_rho(k)*S(:,k)*S(:,k)' + C;
        end
        R = sigma*eye(N);   % Noise Covariance Matrix
        W = (C+R)\s*rho(i);    % W(l) for each l %%finv = (C+R)^-1; f2 = s*rho(i); W2 = finv*f2; W2 = W;
        %%%%%%%%%% Calculate X(l) for Next Step %%%%%%%%%%
        if j == 1
            temp_X(i-N+1) = W'*y_noisy(i:i+N-1);
        else
            temp_X(i-N+1) = W'*y_noisy(i+(N-1)*(j-1):(j-1)*(N-1)+i+N-1);
        end
    end
    X = temp_X(1:length(X)-2*N+2);
%     hold on
%     grid on
end

Recived_profile = X;

File: APC/my_codes/fv/src/general_apc.cpp
#include "../include/general_apc.h"

// additional libraries
#include <chrono>
#include <vector>

General_APC::General_APC()
{

}

cx_mat General_APC::algorithm(cx_mat s, cx_mat  y_noisy, double N, mat alpha, double sigma)
{
	
	// zero padding of y_noisy
	// input y_noisy is 250 * 1 
    double M = alpha.size();
    y_noisy = join_cols(zeros<cx_mat>(M*(N-1),1), y_noisy);
    y_noisy = join_cols(y_noisy, zeros<cx_mat>(M*(N-1),1));
	// output of y_noisy is 298 * 1


    cx_mat S = zeros<cx_mat>(N,2*N-1);
    cx_mat temp = s;
    for (int var = N-1; var < 2 * N - 1; ++var) {
        S.col(var) = temp;
        temp = join_cols(zeros<cx_mat>(1,1), temp.submat(0,0,N-2,0));
    }

    temp = s;
    for (int var = N-2; var >=0; --var) {
        temp = join_cols(temp.submat(1,0,N-1,0),zeros<cx_mat>(1,1));
        S.col(var) = temp;
    }

    cx_mat sum_s = zeros<cx_mat>(N,N);
    for (int var = 0; var < 2*N-1; ++var) {
        sum_s = sum_s + S.col(var) * S.col(var).t();
    }

    cx_mat W_int = inv<cx_mat>(sum_s)*s;
    cx_mat temp_X = zeros<cx_mat>(y_noisy.size()-(N-1),1);

    for (int var = 0; var < y_noisy.size()-(N-1) - 1; ++var) {
        temp_X.row(var) = W_int.st() * y_noisy.submat(var,0,var+N-1,0);
    }

    cx_mat R = eye<cx_mat>(N,N) * sigma;
    cx_mat X = temp_X;

    // Get the starting timepoint
	auto start = std::chrono::high_resolution_clock::now();

    for (int j = 0; j < M; ++j) {									// 2x
        temp_X = zeros<cx_mat>(1,X.size()-2*N+2);
        //cx_mat rho = elementwisePow(X, alpha(j,0));
		arma::mat rho = arma::pow(arma::abs(X), alpha(j, 0)); 

        for (int i = N-1; i < X.size()-N+1; ++i) {					// 262x and 238x
			arma::mat temp_rho = rho.submat(i-N+1,0,i+N-1,0);
            cx_mat C = zeros<cx_mat>(N,N);


			auto time1 = std::chrono::high_resolution_clock::now();
            for (int k = 0; k < 2*N-1; ++k) {						// 25x
                C = temp_rho(k,0)*S.col(k)*S.col(k).t() + C;
            }

			auto time2 = std::chrono::high_resolution_clock::now();
            cx_mat W = inv(C+R)*s*rho(i,0);


			auto time3 = std::chrono::high_resolution_clock::now();
            cx_mat t = W.t() * y_noisy.submat(i+(N-1)*(j),0,(j)*(N-1)+i+N-1,0);


			auto time4 = std::chrono::high_resolution_clock::now();
            temp_X(0, i-N+1) = t(0,0);

        }
        X = temp_X.st();
    }

    return X;
}





// modifying =========================================================================================
cx_mat General_APC::algorithm2(cx_mat s, cx_mat  y_noisy, double N, mat alpha, double sigma)
{


	// zero padding of y_noisy
	// input y_noisy is 250 * 1 
    double M = alpha.size();
    y_noisy = join_cols(zeros<cx_mat>(M*(N-1),1), y_noisy);
    y_noisy = join_cols(y_noisy, zeros<cx_mat>(M*(N-1),1));
	// output of y_noisy is 298 * 1

    cx_mat S = zeros<cx_mat>(N,2*N-1);

    cx_mat temp = s;
    for (int var = N-1; var < 2 * N - 1; ++var) {
        S.col(var) = temp;
        temp = join_cols(zeros<cx_mat>(1,1), temp.submat(0,0,N-2,0));
    }

    temp = s;
    for (int var = N-2; var >=0; --var) {
        temp = join_cols(temp.submat(1,0,N-1,0),zeros<cx_mat>(1,1));
        S.col(var) = temp;
    }

	// modifying ====================================================
	std::vector<arma::cx_mat> Ss(25);                   // size: 25 * 13 * 13
	for(int k = 0; k < 2*N-1; k++) {
        Ss[k] = S.col(k) * S.col(k).t(); 
    }
	// =============================================================


    cx_mat sum_s = zeros<cx_mat>(N,N);                  // size: 13 * 13
    for (int var = 0; var < 2*N-1; ++var) {
        sum_s = sum_s + S.col(var) * S.col(var).t();
    }

    cx_mat W_int = inv<cx_mat>(sum_s)*s;                // size: 13 * 1
    cx_mat temp_X = zeros<cx_mat>(y_noisy.size()-(N-1),1);

    for (int var = 0; var < y_noisy.size()-(N-1) - 1; ++var) {
        temp_X.row(var) = W_int.t() * y_noisy.submat(var,0,var+N-1,0);
    }

    cx_mat R = eye<cx_mat>(N,N) * sigma;
    cx_mat X = temp_X;



	int sample_test = 12 + 0; 


    // Get the starting timepoint
	auto start = std::chrono::high_resolution_clock::now();


	// modifying =============
	cx_mat apc_test; 
	// =======================

    for (int j = 0; j < M; ++j) {									// 2x
        
		// std::cout << "X size: " << X.size() << std::endl; 
		temp_X = zeros<cx_mat>(1,X.size()-2*N+2);
		cx_mat rho = elementwisePow(X, alpha(j,0));
		//apc_test = rho; 


        for (int i = N-1; i < X.size()-N+1; ++i) {  // upper: X.size()-N+1  // 262x and 238x
            //cx_mat temp_rho = rho.submat(i-N+1,0,i+N-1,0);
            cx_mat C = zeros<cx_mat>(N,N);

			// time1
			auto time1 = std::chrono::high_resolution_clock::now();
            for (int k = 0; k < 2*N-1; ++k) {						// 2N-1  25x
                C = rho(i-N+1+k,0)*Ss[k] + C;
            }

			// time2
			auto time2 = std::chrono::high_resolution_clock::now();


            cx_mat W = inv(C+R)*s*rho(i,0);



        }
        X = temp_X.st(); 
    }
    return X;
}
// =====================================================================================================





cx_mat General_APC::elementwisePow(cx_mat input, double p)
{
    cx_mat out = input;
    for (int i = 0; i < input.n_rows; ++i) {
        for (int j = 0; j < input.n_cols; ++j) {
            out(i,j) = pow(input(i,j), p);
        }
    }
    return out;
}
